### PR TITLE
Add X-Vault-Namespace also to auth requests. Would get 400 Bad Reques…

### PR DESCRIPTION
**Pull request by rosuvlad**
Add X-Vault-Namespace also to auth requests. Would get 400 Bad Request - Invalid Role ID otherwise

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@rosuvlad Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-config&pullRequestId=1564)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-cloud/spring-cloud-config&pullRequestId=1564) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@rosuvlad Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-config&pullRequestId=1564)!

**Comment by spencergibb** [comment]
Can you step back and describe the problem you are seeing beyond 
> Would get 400 Bad Request - Invalid Role ID 

What versions are you using?

IIRC X-Vault-Namespace is an enterprise feature that we can not test. @mp911de?

**Comment by spencergibb** [comment]
So I think it was removed here https://github.com/spring-cloud/spring-cloud-config/commit/814d0b1e3953bb865e9bdb1e079f5cb246994fcc#diff-bcc35146178531de2130c6b0069350feL77

**Comment by rosuvlad** [comment]
Using:
 - springCloudVersion: Hoxton.SR1
 - org.springframework.boot: 2.2.4.RELEASE
 - org.springframework.cloud:spring-cloud-config-server: 2.2.1.RELEASE

We are having Enterprise Vault available in our company and we need to use the namespace segregation.

The problem is that the namespace is enforced for Auth endpoints (except the root one, token) in our Vault configuration.

In our case AppRole would be something specific for each namespace and the current state of spring-config-server is not passing X-Vault-Namespace in the calls for the AppRole flow. 
By running the code without the fix, we would encounter a 400 - BadRequest - Invalid Role ID from the Vault backend.

**Comment by rosuvlad** [comment]
> So I think it was removed here [814d0b1#diff-bcc35146178531de2130c6b0069350feL77](https://github.com/spring-cloud/spring-cloud-config/commit/814d0b1e3953bb865e9bdb1e079f5cb246994fcc#diff-bcc35146178531de2130c6b0069350feL77)

Why?

**Comment by spencergibb** [comment]
You've duplicated code found here https://github.com/spring-cloud/spring-cloud-config/blob/5544f03128f5c6194d686d7f9e85303a93439b81/spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/vault/SpringVaultClientConfiguration.java#L98

**Comment by rosuvlad** [comment]
> You've duplicated code found here
> 
> https://github.com/spring-cloud/spring-cloud-config/blob/5544f03128f5c6194d686d7f9e85303a93439b81/spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/vault/SpringVaultClientConfiguration.java#L98

From our debug, the section there is not being called for the AppRole auth flow.

**Comment by rosuvlad** [comment]
> restOperations(

I believe that would be OK and will have the same result.

**Comment by spencergibb** [comment]
Why do you keep closing this? I'm trying to help you figure the right place for this to go to fix your problem. Why did the code get removed? I don't know. We can't check the flow because we don't have vault enterprise to test with, that's why this slipped thru.

**Comment by rosuvlad** [comment]
> Why do you keep closing this? I'm trying to help you figure the right place for this to go to fix your problem. Why did the code get removed? I don't know. We can't check the flow because we don't have vault enterprise to test with, that's why this slipped thru.

Sorry for closing, using Gitlab or Bitbucket on the company premise, not used to GitHub.
So should I create a new PR with the override on restOperations()?

**Comment by mp911de** [comment]
Namespaces are an enterprise feature. In Spring Vault, we [download the enterprise version](https://github.com/spring-projects/spring-vault/blob/master/src/test/bash/install_vault.sh#L139) for testing. The good news is that Vault enterprise works for 30 minutes without a license key. However, [setting up namespaces for testing](https://github.com/spring-projects/spring-vault/blob/master/spring-vault-core/src/test/java/org/springframework/vault/core/VaultNamespaceSecretIntegrationTests.java) isn't trivial.

Namespaces are required to be set for authentication and secret operations. Note that the namespace must not be set for health check operations as these fail when using namespaces.

**Comment by rosuvlad** [comment]
> Namespaces are required to be set for authentication and secret operations. Note that the namespace must not be set for health check operations as these fail when using namespaces.

OK, let me give it another go and also update the tests.
Thank you for the support.

**Comment by rosuvlad** [comment]
OK, updated and tests are passing :)

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/vault/SpringVaultClientConfiguration.java [hunk 1]
```diff
@@ -125,6 +125,18 @@ public SslConfiguration sslConfiguration() {
 		return new SslConfiguration(keyStoreConfiguration, trustStoreConfiguration);
 	}
 
+	@Override
+	public RestOperations restOperations() {
+		RestTemplate restOperations = (RestTemplate) super.restOperations();
+
+		if (vaultProperties.getNamespace() != null) {
```
Comment (spencergibb) [inline spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/vault/SpringVaultClientConfiguration.java]: Let's extract this into a method and also replace https://github.com/spring-cloud/spring-cloud-config/blob/971214645bba8afac71b2e2906a91a36868a0658/spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/vault/SpringVaultClientConfiguration.java#L98-L102

Comment (rosuvlad) [reply spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/vault/SpringVaultClientConfiguration.java]: Sure, doing that now. :)

**Comment by spencergibb** [comment]
#1565 has been opened to track using vault enterprise during CI

**Comment by rosuvlad** [comment]
Will close this PR to retarget to 2.2.x branch

**Pull request closed without merging**